streaming/storage/deque.go

```diff
@@ -0,0 +1,405 @@
+
+	err = dq.tx.Set(byteKey, data)
+	if err != nil {
+		return errors.Wrap(err, "failed to set value in PushBack")
+	}
+
+	err = dq.increaseFirstFreeBackSpot() // we move the free spot at the back to the right
+	if err != nil {
+		return errors.Wrap(err, "failed to increase first free back spot in PushBack")
+	}
+
+	return nil
+}
+
+// Removes the first element from the queue and returns it.
+// Returns ErrNotFound if the list is empty.
+func (dq *Deque) PopFront(value proto.Message) error {
+	// there is no need to call initialize here, since PeekFront calls initialize
+
+	err := dq.PeekFront(value) // retrieve the value
+	if err == ErrNotFound {
+		return ErrNotFound
+	} else if err != nil {
+		return errors.Wrap(err, "failed to read first element in PopFront")
+	}
+
+	// and now we have to remove it from the deque
+
+	key := getIndexKey(dq.firstFreeFrontSpot + 1)
+	err = dq.tx.Delete(key)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete first element in PopFront")
+	}
+
+	err = dq.increaseFirstFreeFrontSpot() // we move the first free front spot to the right
+	if err != nil {
+		return errors.Wrap(err, "failed to modify first free front spot in PopFront")
+	}
+
+	return nil
+}
+
+// Removes the last element from the queue and returns it.
+// Returns ErrNotFound if the list is empty.
+func (dq *Deque) PopBack(value proto.Message) error {
+	// there is no need to call initialize here, since PeekBack calls initialize
+
+	err := dq.PeekBack(value) // retrieve the value
+	if err == ErrNotFound {
+		return ErrNotFound
+	} else if err != nil {
+		return errors.Wrap(err, "failed to read last element in PopBack")
+	}
+
+	// and now we have to remove it from the deque
+
+	key := getIndexKey(dq.firstFreeBackSpot - 1)
+	err = dq.tx.Delete(key)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete last element in PopBack")
+	}
+
+	err = dq.decreaseFirstFreeBackSpot() // we move the first free back spot to the left
+	if err != nil {
+		return errors.Wrap(err, "failed to modify first free back spot in PopBack")
+	}
+
+	return nil
+}
+
+// Returns the first element of the queue without removing it.
+// Returns ErrNotFound if the queue is empty.
+func (dq *Deque) PeekFront(value proto.Message) error {
+	if !dq.initializedFront {
+		err := dq.initializeFront()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in PeekFront")
+		}
+	}
+
+	return dq.getValueAtIndex(value, dq.firstFreeFrontSpot+1)
+}
+
+// Returns the last element of the queue without removing it.
+// Returns ErrNotFound if the queue is empty.
+func (dq *Deque) PeekBack(value proto.Message) error {
+	if !dq.initializedBack {
+		err := dq.initializeBack()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in PeekBack")
+		}
+	}
+
+	return dq.getValueAtIndex(value, dq.firstFreeBackSpot-1)
+}
+
+// Returns the value at given index.
+// We don't call initialize here, since this method is only called in methods like
```

Comment:
I think first and foremost, we don't call initialize here because we don't use any initializable values.

Reply:
We use dq.firstFreeBackSpot

Reply:
We use dq.firstFreeBackSpot